src/main/scala/components/memory-async.scala

```diff
@@ -0,0 +1,141 @@
+  val operation    = UInt(1.W)
+}
+
+// A bundle used for representing the memory's response to a memory port.
+class Response(val blockwidth: Int) extends Bundle {
+  val address      = UInt(32.W)
+  val data         = UInt(blockwidth.W)
+}
+
+/** 
+ * The generic interface for communication between the IMem/DMemPort modules and the backing memory.
+ *
+ * Input:  request, the ready/valid interface for a MemPort module to issue Requests to. Memory
+ *         will only accept a request when both request.valid (the MemPort is supplying valid data)
+ *         and request.ready (the memory is idling for a request) are high.
+ *
+ * Output: response, the valid interface for the data outputted by memory if it was requested to read. 
+ *         the bits in response.bits should only be treated as valid data when response.valid is high.
+ */
+class AsyncMemIO(val blockwidth: Int) extends Bundle {
+  val request  = Flipped(Decoupled (new Request (blockwidth)))
+  val response = Valid (new Response (blockwidth))
+}
+
+/**
+ * The modified asynchronous form of the dual ported memory module. 
+ * When io.imem.request.valid or io.imem.request.valid is true and the memory is ready for an operation,
+ * this memory module simulates the latency of real DRAM by pushing memory accesses into pipes that delay 
+ * the request for a configurable latency.
+ *
+ * As with the synced memory module, this memory should only be instantiated in the Top file,
+ * and never within the actual CPU.
+ *
+ * The I/O for this module is defined in [[AsyncMemIO]].
+ */
+class DualPortedAsyncMemory(size: Int, memfile: String, latency: Int, val blockwidth: Int) extends Module {
+  def wireMemPipe(portio: AsyncMemIO, pipe: Pipe[Request], busy: Bool): Unit = {      
+    pipe.io.enq.bits  <> DontCare
+    pipe.io.enq.valid := false.B
+
+    portio.request.ready := !busy
+  } 
+
+  val io = IO(new Bundle {
+    val imem = new AsyncMemIO(blockwidth)
+    val dmem = new AsyncMemIO(blockwidth)
+  })
+  io <> DontCare
+
+  assert(latency > 0) // Check for attempt to make combinational memory
+
+  val memory    = Mem(math.ceil(size.toDouble/4).toInt, UInt(blockwidth.W))
+  loadMemoryFromFile(memory, memfile)
+
+  // Instruction port
+  val imemPipe = Module(new Pipe(new Request(blockwidth), latency))
+  val imemBusy = RegInit(false.B)
+
+  wireMemPipe(io.imem, imemPipe, imemBusy)
+
+  when (!imemBusy && io.imem.request.valid) {
+    // Put the Request into the instruction pipe and signal that instruction memory is busy
+    val inRequest = io.imem.request.asTypeOf(new Request(blockwidth))
+    imemPipe.io.enq.bits  := inRequest
+    imemPipe.io.enq.valid := true.B
+    imemBusy := true.B
+  }
+
+  when (imemPipe.io.deq.valid) {
+    assert(imemBusy)
+    assert(imemPipe.io.deq.bits.operation === Read) 
+    val outRequest = imemPipe.io.deq.asTypeOf (new Request(blockwidth))
+    io.imem.response.valid        := true.B
+    io.imem.response.bits.data    := memory(outRequest.address >> 2)
```

Comment:
This >> 2 will only work for blockwidth of 4 bytes. Maybe add an assert that the blockwidth is 4 for now?